[CSRanking] Fix func declaration ranking with default'ed parameters

xedin opened the pull request:
If default'ed parameters in one decl are intermixed with non-defaulted
ones, skip claiming parameters in other decl at the same position.

Resolves: rdar://problem/36226874.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@jrose-apple you've added tests which are not ambiguous now, so wanted to get your option on these changes.

DougGregor reviewed (approved):
The changes look like an improvement, but I have to say that I'm bothered by `isDeclAsSpecializedAs` having this logic at all. Can't we re-use the argument-matching logic from the solver?

xedin commented:
@DougGregor Do you mean `matchFunctionTypes` or something else?

xedin commented:
@DougGregor I've changed it to use `matchCallArguments` to determine bindings, I think this is what you meant, is it? (There are couple of tests broken by this, I'll fix them up tomorrow).

jrose-apple commented:
I must have been very nervous when adding the fix-it about not wanting to change ranking too much (see #4785). It looks like I was just being cautious and didn't have any specific fears at the time, so if everyone's good with this then I'm happy to take the improvement!

xedin commented:
Ok, it looks like if we are not matching labels it doesn't really make much sense to use `matchCallArguments` so that's probably way this loop is there now, I'm going to revert to using that.

xedin committed:
[CSRanking] Fix func declaration ranking with default'ed parameters

If default'ed parameters in one decl are intermixed with non-defaulted
ones, skip claiming parameters in other decl at the same position.

Resolves: rdar://problem/36226874.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test macOS platform
